common/src/main/java/com/wynntils/screens/widgets/TextInputBoxWidget.java

```diff
@@ -0,0 +1,215 @@
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.sounds.SoundEvents;
+import org.lwjgl.glfw.GLFW;
+
+// FIXME: This is a very basic text box.
+public class TextInputBoxWidget extends AbstractWidget {
+    protected char defaultCursorChar = '_';
+    protected Consumer<String> onUpdateConsumer;
+    protected String textBoxInput = "";
+    protected int cursorPosition = 0;
+    protected long lastCursorSwitch = 0;
+    protected boolean renderCursor = true;
+
+    protected final WynntilsSettingsScreen settingsScreen;
+
+    public TextInputBoxWidget(
+            int x,
+            int y,
+            int width,
+            int height,
+            Component boxTitle,
+            Consumer<String> onUpdateConsumer,
+            WynntilsSettingsScreen settingsScreen) {
+        super(x, y, width, height, boxTitle);
+        this.onUpdateConsumer = onUpdateConsumer;
+        this.settingsScreen = settingsScreen;
+    }
+
+    public TextInputBoxWidget(
+            int x,
+            int y,
+            int width,
+            int height,
+            Consumer<String> onUpdateConsumer,
+            WynntilsSettingsScreen settingsScreen) {
+        super(x, y, width, height, TextComponent.EMPTY);
+        this.onUpdateConsumer = onUpdateConsumer;
+        this.settingsScreen = settingsScreen;
+    }
+
+    @Override
+    public void render(PoseStack poseStack, int mouseX, int mouseY, float partialTick) {
+        poseStack.pushPose();
+
+        poseStack.translate(this.x, this.y, 0);
+
+        this.renderBg(poseStack, McUtils.mc(), mouseX, mouseY);
+
+        Font font = FontRenderer.getInstance().getFont();
+        final float maxTextWidth = this.width - 4;
+
+        String cursorChar = getRenderCursorChar();
+
+        final float cursorWidth = font.width(String.valueOf(this.getCursorChar()));
+
+        String renderedText;
+        if (font.width(textBoxInput + this.getCursorChar()) < maxTextWidth) {
+            renderedText =
+                    (textBoxInput.substring(0, cursorPosition) + cursorChar + textBoxInput.substring(cursorPosition));
+        } else {
+            StringBuilder builder = new StringBuilder(cursorChar);
+
+            int stringPosition = Math.min(textBoxInput.length() - 1, cursorPosition);
+
+            while (font.width(builder.toString()) < maxTextWidth - cursorWidth && stringPosition >= 0) {
+                builder.append(textBoxInput.charAt(stringPosition));
+
+                stringPosition--;
+            }
+
+            builder.reverse();
+
+            stringPosition = cursorPosition + 1;
+
+            while (font.width(builder.toString()) < maxTextWidth - cursorWidth
+                    && stringPosition < this.textBoxInput.length()) {
+                builder.append(textBoxInput.charAt(stringPosition));
+
+                stringPosition++;
+            }
+
+            renderedText = builder.toString();
+        }
```

Comment:
What does this code do?

Reply:
This is basically to only render text that fits into the text widget (truncate if string is too long). I can add code comments to document this.